lib/validation/bundlers/inputDescriptorsVB.ts

```diff
@@ -9,60 +9,111 @@ export class InputDescriptorsVB extends ValidationBundler<InputDescriptor[]> {
+      | Validation<Constraints>
+      | Validation<Field>
+      | Validation<HolderSubject>
+    )[] = [];
+
+    validations.push(
+      {
+        tag: this.getTag(),
+        target: inputDescriptors,
+        predicate: (inDescs: InputDescriptor[]) => this.shouldHaveUniqueIds(inDescs),
+        message: 'input descriptor ids must be unique',
+      },
+      {
+        tag: this.getTag(),
+        target: inputDescriptors,
+        predicate: (inDescs: InputDescriptor[]) => this.shouldHaveUniqueFieldsIds(inDescs),
+        message: 'fields id must be unique',
+      }
+    );
 
     inputDescriptors.forEach((inputDescriptor, inDescInd) => {
-      validations = [...validations, ...this.getValidationFor(inputDescriptor, inDescInd)];
+      validations = [
+        ...validations,
+        ...this.getValidationFor(inputDescriptor, inDescInd),
+        ...this.constraintsValidations(inputDescriptor, inDescInd),
+      ];
     });
-
-    validations = [
-      ...validations,
-      this.shouldHaveUniqueIds(inputDescriptors),
-      this.shouldHaveUniqueFieldsIds(inputDescriptors),
-    ];
-
     return validations;
   }
 
-  private getValidationFor(inputDescriptor: InputDescriptor, inDescInd: number): Validation<any>[] {
+  private getValidationFor(inputDescriptor: InputDescriptor, inDescInd: number): Validation<InputDescriptor>[] {
     return [
       {
         tag: this.getMyTag(inDescInd),
-        target: inputDescriptor?.id,
-        predicate: InputDescriptorsVB.nonEmptyString,
+        target: inputDescriptor,
+        predicate: (inDesc: InputDescriptor) => InputDescriptorsVB.nonEmptyString(inDesc?.id),
         message: this.idMustBeNonEmptyStringMsg,
       },
       {
         tag: this.getMyTag(inDescInd),
-        target: inputDescriptor?.schema,
+        target: inputDescriptor,
         predicate: this.isValidSchema(),
         message: this.shouldHaveValidSchemaURIMsg,
       },
       {
         tag: this.getMyTag(inDescInd),
-        target: inputDescriptor?.name,
-        predicate: InputDescriptorsVB.optionalNonEmptyString,
+        target: inputDescriptor,
+        predicate: (inDesc: InputDescriptor) => InputDescriptorsVB.optionalNonEmptyString(inDesc?.name),
         message: this.nameShouldBeNonEmptyStringMsg,
       },
       {
         tag: this.getMyTag(inDescInd),
-        target: inputDescriptor?.purpose,
-        predicate: InputDescriptorsVB.optionalNonEmptyString,
+        target: inputDescriptor,
+        predicate: (inDesc: InputDescriptor) => InputDescriptorsVB.optionalNonEmptyString(inDesc?.purpose),
         message: this.purposeShouldBeNonEmptyStringMsg,
       },
-      ...this.constraintsValidations(inputDescriptor, inDescInd),
     ];
   }
 
+  private shouldHaveUniqueFieldsIds(inputDescriptors: InputDescriptor[]): boolean {
+    const nonUniqueInputDescriptorFieldsIds: string[] = [];
+    const uniqueInputDescriptorFieldsIds: Set<string> = new Set<string>();
+    const tmp: Field[] = [];
```

Comment:
Pretty clunky. Why not add the ids to a set and then when encountering the next id check whether the id is present. If this is to shave of a factor of complexity then it is the wrong call given the low amount of iterations that are needed here.

Reply:
I want refactor the validator classes further, I think we should only validate only input descriptors here and not deeply nested properties, just to avoid having to handle multiple types, multiple iteration levels and breaking the single responsibility principle.